ably/realtime_conn.go

```diff
@@ -385,22 +385,47 @@
 			ConnectionStateDisconnected,
 		)
 	}
-	connectMode := c.getMode()
-	query, err := c.params(connectMode)
-	if err != nil {
-		return nil, err
-	}
-	u.RawQuery = query.Encode()
-	proto := c.opts.protocol()
 
-	if c.State() == ConnectionStateClosed { // RTN12d - if connection is closed by client, don't try to reconnect
-		return nopResult, nil
+	var conn conn
+	primaryHost := c.opts.getRealtimeHost()
+	hosts := []string{primaryHost}
+	fallbackHosts, err := c.opts.getFallbackHosts()
+	if err != nil {
+		c.log().Warn(err)
+	} else {
+		hosts = append(hosts, ablyutil.Shuffle(fallbackHosts)...)
 	}
+	// Always try primary host first and then fallback hosts for realtime conn
+	for hostCounter, host := range hosts {
+		u, err := url.Parse(c.opts.realtimeURL(host))
+		if err != nil {
+			return nil, err
+		}
+		query, err := c.params(connectMode)
+		if err != nil {
+			return nil, err
+		}
+		u.RawQuery = query.Encode()
+		proto := c.opts.protocol()
 
-	// if err is nil, raw connection with server is successful
-	conn, err := c.dial(proto, u)
-	if err != nil {
-		return nil, err
+		if c.State() == ConnectionStateClosed { // RTN12d - if connection is closed by client, don't try to reconnect
+			return nopResult, nil
+		}
+		// if err is nil, raw connection with server is successful
+		conn, err = c.dial(proto, u)
+		if err != nil {
+			resp := extractHttpResponseFromError(err)
+			if hostCounter < len(hosts)-1 && canFallBack(err, resp) && c.opts.hasActiveInternetConnection() { // RTN17d, RTN17c
+				continue
+			}
+			return nil, err
+		}
+		if host != primaryHost { // RTN17e
+			c.client.rest.setActiveRealtimeHost(host)
+		} else if !empty(c.client.rest.activeRealtimeHost) {
+			c.client.rest.setActiveRealtimeHost("") // reset to default
+		}
+		break
```

Comment:
This is wrong code snippet. We only set active realtime host when host is fallback host. I.e. not primary host. Also, I'm not sure why hostcounter check is removed. It is there to make sure we return error for last iteration. I.e. since next host is not available